evidence/reactor.go

```diff
@@ -21,233 +46,346 @@ const (
-	return evR
+
+	evpool.SetLogger(logger)
+
+	r.BaseService = *service.NewBaseService(logger, "Evidence", r)
+	return r
 }
 
-// SetLogger sets the Logger on the reactor and the underlying Evidence.
-func (evR *Reactor) SetLogger(l log.Logger) {
-	evR.Logger = l
-	evR.evpool.SetLogger(l)
+// SetEventBus implements events.Eventable.
+func (r *Reactor) SetEventBus(b *types.EventBus) {
+	r.eventBus = b
 }
 
-// GetChannels implements Reactor.
-// It returns the list of channels for this reactor.
-func (evR *Reactor) GetChannels() []*p2p.ChannelDescriptor {
-	return []*p2p.ChannelDescriptor{
-		{
-			ID:                  EvidenceChannel,
-			Priority:            5,
-			RecvMessageCapacity: maxMsgSize,
-		},
-	}
+// OnStart starts separate go routines for each p2p Channel and listens for
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed. No error is returned.
+func (r *Reactor) OnStart() error {
+	go r.processEvidenceCh()
+	go r.processPeerUpdates()
+
+	return nil
 }
 
-// AddPeer implements Reactor.
-func (evR *Reactor) AddPeer(peer p2p.Peer) {
-	go evR.broadcastEvidenceRoutine(peer)
+// OnStop stops the reactor by signaling to all spawned goroutines to exit and
+// blocking until they all exit.
+func (r *Reactor) OnStop() {
+	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
+	// p2p Channels should execute Close().
+	close(r.closeCh)
+
+	// Wait for all spawned peer evidence broadcasting goroutines to gracefully
+	// exit.
+	r.peerWG.Wait()
+
+	// Wait for all p2p Channels to be closed before returning. This ensures we
+	// can easily reason about synchronization of all p2p Channels and ensure no
+	// panics will occur.
+	<-r.evidenceCh.Done()
+	<-r.peerUpdates.Done()
 }
 
-// Receive implements Reactor.
-// It adds any received evidence to the evpool.
-// XXX: do not call any methods that can block or incur heavy processing.
-// https://github.com/tendermint/tendermint/issues/2888
-func (evR *Reactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
-	evis, err := decodeMsg(msgBytes)
-	if err != nil {
-		evR.Logger.Error("Error decoding message", "src", src, "chId", chID, "err", err)
-		evR.Switch.StopPeerForError(src, err)
-		return
+// handleEvidenceMessage handles enevelopes sent from peers on the EvidenceChannel.
+// It returns an error only if the Envelope.Message is unknown for this channel
+// or if the given evidence is invalid. This should never be called outside of
+// handleMessage.
+func (r *Reactor) handleEvidenceMessage(envelope p2p.Envelope) error {
+	switch msg := envelope.Message.(type) {
+	case *tmproto.EvidenceList:
+		r.Logger.Debug("received evidence list", "num_evidence", len(msg.Evidence), "peer", envelope.From.String())
+
+		// TODO: Refactor the Evidence type to not contain a list since we only ever
```

Comment:
do we have an issue for this?

Reply:
Once upon a time there was this: https://github.com/tendermint/tendermint/issues/4729

But I think following this PR, we might want to jot down all the potential improvements/changes that could be made

Reply:
Why was that closed @cmwaters? Would it still be beneficial to reference that issue in the code?

Reply:
I believe we decided to go initially with individual evidence and then later we decided to generalise light client attack evidence and in doing so batch it together (although duplicate vote evidence is still individual).

Maybe we want to revisit this at a later point.

Reply:
Ok, I will still reference it then so future readers can have context.